File: HW10/HW10.1/Stack.cpp
#include "Stack.h"

void push(Stack& stack, Node* node)
{
	if (!&stack)
	{
		return;
	}
	StackElement* newElement = new StackElement{ stack.head, node };
	stack.head = newElement;
	return;
}

StackElement* pop(Stack& stack)
{
	if (&stack && !isEmpty(stack))
	{
		StackElement* topElement = stack.head;
		stack.head = stack.head->next;
		return topElement;
	}
	return nullptr;
}

void deleteStack(Stack& stack)
{
	if (!&stack)
	{
		return;
	}
	while (stack.head)
	{
		StackElement* temp = stack.head->next;
		if (stack.head->node)
		{
			delete stack.head->node;
		}
		delete stack.head;
		stack.head = temp;
	}
	return;
}

void deleteElement(StackElement& element)
{
	if (element.node)
	{
		delete element.node;
	}
	delete &element;
	return;
}

bool isEmpty(const Stack& stack)
{
	if (&stack && stack.head)
	{
		return false;
	}
	return true;
}

File: java/HW1/HW1.1/com/spbu/group144/baranovsky/stack/Stack.java
package com.spbu.group144.baranovsky.stack;

/**
 * A stack for storing integer values.
 */
public class Stack {
    private StackElement head;

    public Stack(){
    }

    /**
     * Removes an element from the top of the stack.
     * @return Returns the integer value of the top element.
     */
    public int pop(){
        if(!isEmpty()){
            StackElement top = head;
            head = head.getNext();
            return top.getValue();
        }
        return 0;
    }

    /**
     * Adds a new element to the stack.
     * @param value The integer value of the new element.
     */
    public void push(int value){
        head = new StackElement(value, head);
    }

    /**
     * Checks whether the stack is empty.
     * @return True if the stack is empty. False otherwise.
     */
    public boolean isEmpty(){
        return head == null;
    }
}
